src/main/java/org/folio/oaipmh/helpers/GetOaiMetadataFormatsHelper.java

```diff
@@ -0,0 +1,142 @@
+    this.httpClient = httpClient;
+    this.okapiHeaders = okapiHeaders;
+    this.ctx = ctx;
+  }
+
+  @Override
+  public CompletableFuture<String> handle(Request request, Context ctx) {
+    String identifier = request.getIdentifier();
+    if(identifier!=null) {
+      return retrieveMetadataFormats(identifier);
+    } else {
+      return retrieveMetadataFormats();
+    }
+  }
+
+  /**
+   * Processes request with identifier
+   * @return future with {@link OAIPMH} response
+   */
+  private CompletableFuture<String> retrieveMetadataFormats(String identifier) {
+    CompletableFuture<String> future = new VertxCompletableFuture<>(ctx);
+    try {
+      httpClient.request(HttpMethod.GET, "/instance-storage/instances/" + identifier, okapiHeaders)
+        .thenApply(this::verifyAndGetOaiPmhResponse)
+        .thenAccept(oaipmh -> {
+          try {
+            String reply = ResponseHelper.getInstance().writeToString(oaipmh);
+            future.complete(reply);
+          } catch (JAXBException e) {
+            logger.error("Error happened in OAI-PMH response marshalling", e);
+          }
+        })
+        .exceptionally(t -> {
+          future.completeExceptionally(t);
+          return null;
+        });
+    } catch (Exception e) {
+      future.completeExceptionally(e);
+    }
+    return future;
+  }
+
+  /**
+   * Processes request without identifier
+   * @return future with {@link OAIPMH} response
+   */
+  private CompletableFuture<String> retrieveMetadataFormats() {
+    CompletableFuture<String> future = new VertxCompletableFuture<>(ctx);
+    try {
+      future
+        .complete(ResponseHelper.getInstance().writeToString(buildMetadataFormatTypesResponse()));
+    } catch (Exception e) {
+      logger.error("Error happened while processing ListMetadataFormats verb request", e);
+      future.completeExceptionally(e);
+    }
+    return future;
+  }
+
+  /**
+   * Validates inventory-mod-storage response and returns {@link OAIPMH}
+   * with populated MetadataFormatTypes or needed Errors according to OAI-PMH2 specification
+   * @return basic {@link OAIPMH}
+   */
+  private OAIPMH verifyAndGetOaiPmhResponse(Response response) {
+    if (!Response.isSuccess(response.getCode())) {
+      return buildIdentifierNotFound();
+    }
+    return buildMetadataFormatTypesResponse();
+  }
+
+  /**
+   * Creates {@link OAIPMH} with ListMetadataFormats element
+   * @return basic {@link OAIPMH}
+   */
+  private OAIPMH buildMetadataFormatTypesResponse() {
+    return objectFactory.createOAIPMH()
+      .withListMetadataFormats(getMetadataFormatTypes())
+      .withResponseDate(Instant.now().truncatedTo(ChronoUnit.SECONDS))
+      .withRequest(objectFactory.createRequestType()
+        .withVerb(VerbType.LIST_METADATA_FORMATS));
+  }
+
+  /**
+   * Creates {@link OAIPMH} with Error id-does-not-exist
+   * @return basic {@link OAIPMH}
+   */
+  private OAIPMH buildIdentifierNotFound() {
+    return objectFactory.createOAIPMH()
```

Comment:
It's easier and less verbose to create JAXB object via constructor.
Not necessary to change it now, but in future I would suggest to stick to constructors.

Reply:
Noted but without code changes now.
Thanks!